File: jihoon/Coding_test (1)/Algorithms/Sorting_alg/two_elements_mergesort/two_elem_mergesort.py
def compare_tuples(t1, t2):
    """Custom comparison function for tuples"""
    if t1[1] < t2[1]:  # compare second element
        return -1
    elif t1[1] > t2[1]:
        return 1
    else:  # If second elements are equal, compare the first elements
        if t1[0] < t2[0]:
            return -1
        elif t1[0] > t2[0]:
            return 1
        else:
            return 0

def merge(L, first, mid, last):
    k = first
    sub1 = L[first:mid+1]
    sub2 = L[mid+1:last+1]
    i = j = 0
    while i < len(sub1) and j < len(sub2):
        if compare_tuples(sub1[i], sub2[j]) <= 0:
            L[k] = sub1[i]
            i = i + 1
        else:
            L[k] = sub2[j]
            j = j + 1
        k = k + 1
    if i < len(sub1):
        L[k:last+1] = sub1[i:]
    elif j < len(sub2):
        L[k:last+1] = sub2[j:]

def mergeSortHelp(L, first, last):
    if first == last:
        return
    else:
        mid = first + (last - first) // 2
        mergeSortHelp(L, first, mid)
        mergeSortHelp(L, mid+1, last)
        merge(L, first, mid, last)

def mergeSort(L):
    mergeSortHelp(L, 0, len(L)-1)

if __name__ == '__main__':
    a = [(2, 3), (1, 3), (1, 2), (4, 2), (5, 6), (3, 6)]
    mergeSort(a)
    print(a)

File: jihoon/Coding_test (1)/Algorithms/Sorting_alg/two_elements_mergesort/two_elem_mergesort.cpp
#include <iostream>
#include <vector>

using namespace std;

bool compare_pairs(const pair<int, int>& p1, const pair<int, int>& p2) {
    if (p1.second < p2.second) {
        return true;
    } else if (p1.second == p2.second) {
        return p1.first < p2.first;
    } else {
        return false;
    }
}

void merge(vector<pair<int, int>>& v, int first, int mid, int last) {
    int i = first, j = mid + 1, k = 0;
    vector<pair<int, int>> temp(last - first + 1);

    while (i <= mid && j <= last) {
        if (compare_pairs(v[i], v[j])) {
            temp[k++] = v[i++];
        } else {
            temp[k++] = v[j++];
        }
    }
    
    while (i <= mid) {
        temp[k++] = v[i++];
    }
    
    while (j <= last) {
        temp[k++] = v[j++];
    }
    
    for (i = first, k = 0; i <= last; ++i, ++k) {
        v[i] = temp[k];
    }
}

void mergeSortHelper(vector<pair<int, int>>& v, int first, int last) {
    if (first < last) {
        int mid = first + (last - first) / 2;
        mergeSortHelper(v, first, mid);
        mergeSortHelper(v, mid + 1, last);
        merge(v, first, mid, last);
    }
}

void mergeSort(vector<pair<int, int>>& v) {
    mergeSortHelper(v, 0, v.size() - 1);
}

int main() {
    vector<pair<int, int>> a = {{2, 3}, {1, 3}, {1, 2}, {4, 2}, {5, 6}, {3, 6}};
    mergeSort(a);

    for (const auto& p : a) {
        cout << "(" << p.first << ", " << p.second << ") ";
    }
    cout << endl;

    return 0;
}